thoth/adviser/prescription/v1/pydantic.py

```diff
@@ -0,0 +1,108 @@
+# thoth-adviser
+# Copyright(C) 2023 Max Gautier
+#
+# This program is free software: you can redistribute it and / or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""Prescription pydantic models."""
+
+from typing import List, Union, Optional, Dict, Type, Generic, TypeVar
+
+from pydantic import BaseModel, Field, conint, constr, conlist, validator, create_model
+from pydantic.generics import GenericModel
+from packaging.specifiers import SpecifierSet, InvalidSpecifier
+from thoth.adviser.enums import RecommendationType, DecisionType
+from thoth.adviser.cpu_db import CPUDatabase
+
+_string = constr(min_length=1)
+ModelT = TypeVar('ModelT')
+
+
+def _list(type: Type) -> Type:
+    return conlist(type=type, min_items=1)
+
+_optional_integers = conlist(type=Union[int, None], min_items=1)
+
+class Not(GenericModel, Generic[ModelT]):
```

Comment:
I'm confused about what `Not` is, what does it represent?

Reply:
It's a parametric type. `Not[List[int]]` is 

```python
class NotListInt:
    not: List[int]
```
Which is intented for some parts of the schema which can be either something, or a single-key dict ("not") mapping to somehting (for example https://github.com/thoth-station/adviser/blob/master/thoth/adviser/prescription/v1/schema.py#L106-L114 )

We would use it as :
```python
cpu_families: Union[Not[CpuFamiliesListType], CpuFamiliesListType]
```
in yaml that would be :
```yaml
cpu_families:
- 5
- 9
# and
cpu_families:
  not:
  - 8
  - 9
# are both valid
```

there is further complications because `not` is a reserved word and can't be a attribute, but that's orthogonal

Reply:
Got it. This makes sense, it is just a bit confusing semantically. My first instinct was that:
```python
Union[Not[CpuFamiliesListType], CpuFamilliesListType]
```
was just equivalent to:
```python
typing.Any
```
Thanks for the explanation